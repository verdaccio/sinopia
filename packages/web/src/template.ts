import { getManifestValue, Manifest } from './manifest';

export type TemplateUIOptions = {
  title?: string;
  uri?: string;
  darkMode?: boolean;
  protocol?: string;
  host?: string;
  url_prefix?: string;
  base?: string;
  primaryColor?: string;
  version?: string;
  logoURI?: string;
  scope?: string;
  language?: string;
};

export type Template = {
  manifest: Manifest;
  options: TemplateUIOptions;
  scriptsBodyAfter?: string[];
  metaScripts?: string[];
  bodyBefore?: string;
  bodyAfter?: string;
};

// the outcome of the Webpack Manifest Plugin
export interface WebpackManifest {
  [key: string]: string;
}

export default (template: Template, manifest: WebpackManifest) => `
    <!DOCTYPE html>
      <html lang="en-us"> 
      <head>
        <meta charset="utf-8">
        <title>${template?.options?.title ?? ''}</title>        
        <link rel="icon" type="image/png" href="${template.manifest.ico}" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        ${getManifestValue(template.manifest.css, manifest).map(
          (item) => ` <link href="${item}" rel="stylesheet">`
        )}        
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS=${JSON.stringify(template.options)}
        </script>
        ${template.metaScripts ? template.metaScripts.map((item) => item) : ''}
      </head>    
      <body class="body">
        ${template.bodyBefore ? template.bodyBefore : ''}
        <div id="root"></div>
        ${getManifestValue(template.manifest.js, manifest).map(
          (item) => `<script defer="defer" src="${item}"></script>`
        )}
        ${template.scriptsBodyAfter ? template.scriptsBodyAfter.map((item) => item) : ''}
      </body>
    </html>
`;
