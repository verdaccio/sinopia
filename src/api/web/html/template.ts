import buildDebug from 'debug';
import { getManifestValue, Manifest } from './manifest';

const debug = buildDebug('verdaccio');

export type TemplateUIOptions = {
  title?: string;
  uri?: string;
  darkMode?: boolean;
  protocol?: string;
  host?: string;
  url_prefix?: string;
  base: string;
  primaryColor?: string;
  version?: string;
  logoURI?: string;
  scope?: string;
  language?: string;
};

export type Template = {
  manifest: Manifest;
  options: TemplateUIOptions;
  scriptsBodyAfter?: string[];
  metaScripts?: string[];
  bodyBefore?: string[];
};

// the outcome of the Webpack Manifest Plugin
export interface WebpackManifest {
  [key: string]: string;
}

export default function renderTemplate(template: Template, manifest: WebpackManifest) {
  debug('template %o', template);
  debug('manifest %o', manifest);
  return `
    <!DOCTYPE html>
      <html lang="en-us"> 
      <head>
        <meta charset="utf-8">
        <base href="${template?.options.base}">
        <title>${template?.options?.title ?? ''}</title>        
        <link rel="icon" href="${template?.options.base}-/static/favicon.ico"/>
        <meta name="viewport" content="width=device-width, initial-scale=1" /> 
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS=${JSON.stringify(template.options)}
        </script>
        ${template?.metaScripts ? template.metaScripts.map((item) => item) : ''}
      </head>    
      <body class="body">
      ${template?.bodyBefore ? template.bodyBefore.map((item) => item) : ''}
        <div id="root"></div>
        ${getManifestValue(template.manifest.js, manifest, template?.options.base).map((item) => `<script defer="defer" src="${item}"></script>`)}
        ${template?.scriptsBodyAfter ? template.scriptsBodyAfter.map((item) => item) : ''}
      </body>
    </html>
  `;
}
