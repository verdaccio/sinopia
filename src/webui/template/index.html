<!DOCTYPE html>
<html lang="en-us">
<head>
    <meta charset="utf-8">
    <title><%= htmlWebpackPlugin.options.title %></title>
    <link rel="icon" type="image/png" href="<%= htmlWebpackPlugin.options.verdaccioURL %>/-/static/favicon.ico"/>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <script>
        window.VERDACCIO_API_URL = '<%= htmlWebpackPlugin.options.verdaccioURL %>/-/verdaccio/';
        window.VERDACCIO_SCOPE = '<%= htmlWebpackPlugin.options.scope %>';
    </script>
</head>
<body class="body">
    <div id="root"></div>
</body>
</html>
