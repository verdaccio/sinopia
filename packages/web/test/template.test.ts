import template from '../src/template';
const manifest = require('./partials/manifest/manifest.json');

const exampleManifest = {
  css: ['main.css'],
  js: ['runtime.js', 'main.js'],
  ico: '/static/foo.ico',
};

describe('template', () => {
  test('custom render', () => {
    expect(template({ options: {}, manifest: exampleManifest }, manifest)).toMatchSnapshot();
  });

  test('custom title', () => {
    expect(
      template({ options: { title: 'foo title' }, manifest: exampleManifest }, manifest)
    ).toMatchSnapshot();
  });

  test('custom body after', () => {
    expect(
      template({ bodyAfter: [`<script src="foo"/>`], manifest: exampleManifest }, manifest)
    ).toMatchSnapshot();
  });

  test('custom body before', () => {
    expect(
      template(
        {
          bodyBefore: [`<script src="fooBefore"/>`, `<script src="barBefore"/>`],
          manifest: exampleManifest,
        },
        manifest
      )
    ).toMatchSnapshot();
  });
});
