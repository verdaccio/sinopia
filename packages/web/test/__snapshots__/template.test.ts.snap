// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template custom body after 1`] = `
"
    <!DOCTYPE html>
      <html lang=\\"en-us\\"> 
      <head>
        <meta charset=\\"utf-8\\">
        <title></title>        
        <link rel=\\"icon\\" type=\\"image/png\\" href=\\"/static/foo.ico\\" />
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
         <link href=\\"/-/static/main.6f2f2cccce0c813b509f.css\\" rel=\\"stylesheet\\">        
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS=undefined
        </script>
        
      </head>    
      <body class=\\"body\\">
        
        <div id=\\"root\\"></div>
        <script defer=\\"defer\\" src=\\"/-/static/runtime.6126058572f989c948b1.js\\"></script>,<script defer=\\"defer\\" src=\\"/-/static/main.6126058572f989c948b1.js\\"></script>
        
      </body>
    </html>
"
`;

exports[`template custom body before 1`] = `
"
    <!DOCTYPE html>
      <html lang=\\"en-us\\"> 
      <head>
        <meta charset=\\"utf-8\\">
        <title></title>        
        <link rel=\\"icon\\" type=\\"image/png\\" href=\\"/static/foo.ico\\" />
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
         <link href=\\"/-/static/main.6f2f2cccce0c813b509f.css\\" rel=\\"stylesheet\\">        
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS=undefined
        </script>
        
      </head>    
      <body class=\\"body\\">
        <script src=\\"fooBefore\\"/>,<script src=\\"barBefore\\"/>
        <div id=\\"root\\"></div>
        <script defer=\\"defer\\" src=\\"/-/static/runtime.6126058572f989c948b1.js\\"></script>,<script defer=\\"defer\\" src=\\"/-/static/main.6126058572f989c948b1.js\\"></script>
        
      </body>
    </html>
"
`;

exports[`template custom render 1`] = `
"
    <!DOCTYPE html>
      <html lang=\\"en-us\\"> 
      <head>
        <meta charset=\\"utf-8\\">
        <title></title>        
        <link rel=\\"icon\\" type=\\"image/png\\" href=\\"/static/foo.ico\\" />
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
         <link href=\\"/-/static/main.6f2f2cccce0c813b509f.css\\" rel=\\"stylesheet\\">        
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS={}
        </script>
        
      </head>    
      <body class=\\"body\\">
        
        <div id=\\"root\\"></div>
        <script defer=\\"defer\\" src=\\"/-/static/runtime.6126058572f989c948b1.js\\"></script>,<script defer=\\"defer\\" src=\\"/-/static/main.6126058572f989c948b1.js\\"></script>
        
      </body>
    </html>
"
`;

exports[`template custom title 1`] = `
"
    <!DOCTYPE html>
      <html lang=\\"en-us\\"> 
      <head>
        <meta charset=\\"utf-8\\">
        <title>foo title</title>        
        <link rel=\\"icon\\" type=\\"image/png\\" href=\\"/static/foo.ico\\" />
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
         <link href=\\"/-/static/main.6f2f2cccce0c813b509f.css\\" rel=\\"stylesheet\\">        
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS={\\"title\\":\\"foo title\\"}
        </script>
        
      </head>    
      <body class=\\"body\\">
        
        <div id=\\"root\\"></div>
        <script defer=\\"defer\\" src=\\"/-/static/runtime.6126058572f989c948b1.js\\"></script>,<script defer=\\"defer\\" src=\\"/-/static/main.6126058572f989c948b1.js\\"></script>
        
      </body>
    </html>
"
`;
